### PR TITLE
Add states

### LINKED ISSUE #6: Use case for states
String interpolation. It's hard to tokenize this otherwise:

```js
a.b(`c \` ${d.e(`f`, g => {})} h \x0a`)
```

It's also nice to let states inherit rules from other states and be able to count things, e.g. the interpolated expression state needs a `}` rule that can tell if it's a closing brace or the end of the interpolation, but is otherwise identical to the normal expression state.

### PR DESCRIPTION
**Pull request by nathan**
Closes #6.

I avoided using Maps, ~~but they would be~~ and they are not actually much faster.

### CODE REVIEW

**Comment by tjvr** [comment]
> I avoided using Maps, but they would be faster.

If they're noticeably faster, go for it. :-)

**Comment by nathan** [comment]
> If they're noticeably faster, go for it. :-)

Did some benchmarks; they're not.

File: test/test.js [hunk 1]
```diff
@@ -142,6 +165,63 @@ describe('moo lexer', () => {
 })
 
 
+describe('moo stateful lexer', () => {
```
Comment (tjvr) [inline test/test.js]: Can we rename this and the `'states'` tests to be more clearly different?

File: moo.js [hunk 2]
```diff
@@ -229,6 +284,10 @@
       col: this.col,
     }
 
+    if (group.pop) this.popState()
+    else if (group.push) this.pushState(group.push)
+    else if (group.next) this.setState(group.next)
```
Comment (tjvr) [inline moo.js]: minor: prefer braces here

File: moo.js [hunk 3]
```diff
@@ -154,6 +190,25 @@
     lineBreaks: true,
   }
 
+  Lexer.prototype.setState = function(state) {
+    if (!state || this.state === state) return
+    this.state = state
+    var index = this.re ? this.re.lastIndex : 0
+    var info = this.states[state]
```
Comment (tjvr) [inline moo.js]: Suggestion: warn if state is not known.

Also, wouldn't it make sense for `lexer.setState()` to be an error?

File: moo.js [hunk 4]
```diff
@@ -138,14 +144,44 @@
...
+  function compile(rules) {
+    var result = compileRules(rules)
+    return new Lexer({start: result}, 'start')
+  }
+
+  function compileStates(states, start) {
+    var keys = Object.getOwnPropertyNames(states)
+    if (!start) start = keys[0]
+
+    var map = {}
```
Comment (tjvr) [inline moo.js]: Is Object.create(null) better here?

File: test/test.js (hunk 1 above)
Comment (nathan) [inline test/test.js]: Not sure what you mean by that.

File: moo.js (hunk 4 above)
Comment (nathan) [inline moo.js]: It doesn't make a difference behavior-wise, but I think it is a bit clearer to use `Object.create(null)`

File: moo.js (hunk 3 above)
Comment (nathan) [inline moo.js]: > Suggestion: warn if state is not known.

We do this at compile-time. For performance, the fewer things you do at runtime, the better.

File: moo.js (hunk 2 above)
Comment (tjvr) [inline moo.js]: Sheesh, that actually does look 100% worse. Let's drop that commit. :P

File: moo.js (hunk 3 above)
Comment (tjvr) [inline moo.js]: Oh, neat. Yup, I agree with all of the above :)

**Comment by tjvr** [comment]
Ugh, turns out I can't merge this from the command-line *and* drop the braces commit. :P

Anyway, this is great! Merge when ready. :)

**Comment by nathan** [comment]
Did you force-push `master` at some point? #17 thinks https://github.com/tjvr/moo/commit/178d6ec50f6a7d3d6b0354ed3665af5b54f9324c should be in `master`, but it's not

**Comment by tjvr** [comment]
> Did you force-push master at some point? 

I tried to merge this (#19) sans 2718735. GitHub didn't like it, so I undid it. Sorry!

**Comment by nathan** [comment]
No problem. You can just revert https://github.com/tjvr/moo/commit/271873530c41888cf05f546bf3bdedfe0c40dfaf if you want.

**Pull request merged**